agent/taskresource/fsxwindowsfileserver/fsxwindowsfileserver_windows.go

```diff
@@ -0,0 +1,663 @@
+			return err
+		}
+	} else {
+		err = errors.New("unsupported credentialsParameter ARN, only ssm/secretsmanager ARNs are valid")
+		fv.setTerminalReason(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (fv *FSxWindowsFileServerResource) retrieveSSMCredentials(credentialsParameterARN string, iamCredentials credentials.IAMRoleCredentials) error {
+	parsedARN, err := arn.Parse(credentialsParameterARN)
+	if err != nil {
+		fv.setTerminalReason(err.Error())
+		return err
+	}
+
+	ssmClient := fv.ssmClientCreator.NewSSMClient(fv.region, iamCredentials)
+	ssmParam := filepath.Base(parsedARN.Resource)
+	ssmParams := []string{ssmParam}
+
+	ssmParamMap, err := ssm.GetParametersFromSSM(ssmParams, ssmClient)
+	if err != nil {
+		fv.setTerminalReason(err.Error())
+		return err
+	}
+
+	ssmParamData, _ := ssmParamMap[ssmParam]
+	creds := FSxWindowsFileServerCredentials{}
+
+	if err := json.Unmarshal([]byte(ssmParamData), &creds); err != nil {
+		return err
+	}
+
+	fv.SetCredentials(creds)
+	return nil
+}
+
+func (fv *FSxWindowsFileServerResource) retrieveASMCredentials(credentialsParameterARN string, iamCredentials credentials.IAMRoleCredentials) error {
+	_, err := arn.Parse(credentialsParameterARN)
+	if err != nil {
+		fv.setTerminalReason(err.Error())
+		return err
+	}
+
+	asmClient := fv.asmClientCreator.NewASMClient(fv.region, iamCredentials)
+	asmData, err := asm.GetSecretFromASM(credentialsParameterARN, asmClient)
+	if err != nil {
+		return err
+	}
+
+	creds := FSxWindowsFileServerCredentials{}
+	if err := json.Unmarshal([]byte(asmData), &creds); err != nil {
+		return err
+	}
+
+	fv.SetCredentials(creds)
+	return nil
+}
+
+func (fv *FSxWindowsFileServerResource) retrieveFileSystemDNSName(fileSystemId string, iamCredentials credentials.IAMRoleCredentials) error {
+	fileSystemIds := []string{fileSystemId}
+	fsxClient := fv.fsxClientCreator.NewFSxClient(fv.region, iamCredentials)
+	fileSystemDNSMap, err := fsx.GetFileSystemDNSNames(fileSystemIds, fsxClient)
+	if err != nil {
+		fv.setTerminalReason(err.Error())
+		return err
+	}
+
+	fv.SetFileSystemDNSName(fileSystemDNSMap[fileSystemId])
+	return nil
+}
+
+var execCommand = exec.Command
+
+func (fv *FSxWindowsFileServerResource) performHostMount(remotePath string, localPath string, username string, password string) error {
+	// formatting to keep powershell happy
+	localPath = fmt.Sprintf("-LocalPath %s", localPath)
+	remotePath = fmt.Sprintf("-RemotePath %s", remotePath)
+	creds := fmt.Sprintf("-Credential $(New-Object System.Management.Automation.PSCredential(%s, $(ConvertTo-SecureString %s -AsPlainText -Force)))", username, password)
+
+	cmd := execCommand("powershell.exe",
+		"New-SmbGlobalMapping",
+		localPath,
+		remotePath,
+		creds,
+		"-Persistent $true",
+		"-RequirePrivacy $true",
+		"-ErrorAction Stop")
+
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		fv.setTerminalReason(err.Error())
+		return err
+	}
+
+	cmd = execCommand("powershell.exe",
```

Comment:
What happens if the agent crashes mid-way during this method?

Reply:
good question.

Reply:
This is a valid concern! thanks for bringing this up @vsiddharth . Here's a thought - I am thinking to introduce an additional check here to verify if the allocated drive letter is still available. In case the agent crashes after the resource was created (but the resource status wasn't updated), this check would fail and would simply return the err as drive occupied.

Reply:
what if the creds change after agent restarts? do we plan to pull in the creds from SSM/ASM always?

Reply:
discussed with @sharanyad offline. before doing a mount, we'll check if the drive letter is available. if not, we'll remove and remount it since the ```New-SmbGlobalMapping``` powershell cmdlet always needs a unoccupied drive letter.